pierone/api.py

```diff
@@ -26,35 +28,97 @@ def __str__(self):
-        return DockerImage(registry=registry, team=team, artifact=artifact, tag=tag)
-
-    def __str__(self):
-        '''
-        >>> str(DockerImage(registry='registry', team='foo', artifact='bar', tag='1.9'))
-        'registry/foo/bar:1.9'
-        '''
-        return '{}/{}/{}:{}'.format(*tuple(self))
+class PierOne:
+
+    def __init__(self, url: str):
+        self.url = url if url.startswith("https://") else "https://" + url
+        self.__access_token = get_token('pierone', ['uid'])
+        self.session = requests.Session()
+        self.session.headers['Authorization'] = 'Bearer {}'.format(self.__access_token)
+
+    def _get(self, path, *args, **kwargs) -> requests.Response:
+        """
+        GETs things from Pier One.
+        """
+        url = self.url + path
+        response = self.session.get(url, *args, **kwargs)
+        response.raise_for_status()
+        return response
+
+    def _post(self, path, json=None, *args, **kwargs) -> requests.Response:
+        """
+        POSTs things to Pier One.
+        """
+        url = self.url + path
+        response = self.session.post(url, json=json, *args, **kwargs)
+        response.raise_for_status()
+        return response
+
+    def get_tag_info(self, image: DockerImage) -> list:
+        """
+        Gets detailed tag information
+        """
+        path = "/teams/{}/artifacts/{}/tags/{}".format(image.team, image.artifact, image.tag)
+
+        try:
+            response = self._get(path)
+        except requests.HTTPError as error:
+            if error.response.status_code == 404:
+                # Raise ArtifactNotFound only if it the image is not found
+                raise ArtifactNotFound(image)
+            raise
+        tag_info = response.json()
+        created_by = tag_info["created_by"]
+        tag_info["created_by"] = KNOWN_USERS.get(created_by, created_by)
+        return tag_info
+
+    def get_image_tags(self, image: DockerImage) -> list:
+        """
+        Gets all tags for an image.
+        """
+        path = "/teams/{team}/artifacts/{artifact}/tags".format(team=image.team, artifact=image.artifact)
+
+        try:
+            response = self._get(path)
+        except requests.HTTPError as error:
+            if error.response.status_code == 404:
+                # Raise ArtifactNotFound only if it the team or artifact are not found
+                raise ArtifactNotFound(image)
+            raise
+        return [parse_pierone_artifact_dict(entry, image.team, image.artifact)
+                for entry in response.json()]
+
+    def get_scm_source(self, image: DockerImage) -> dict:
+        """
+        GETs ``image``s scm_source
+        """
+        path = "/teams/{}/artifacts/{}/tags/{}/scm-source".format(image.team, image.artifact, image.tag)
+        try:
+            response = self._get(path)
+        except requests.HTTPError as error:
+            if error.response.status_code == 404:  # Only return ``None``` if image is not found
```

Comment:
I see the same handling of `404` three times in a row. Maybe it would be better to factor it to `_get_artifact_or_raise` or similar and use it instead?